readthedocs/rtd_tests/tests/test_views.py

```diff
@@ -294,5 +294,15 @@ def test_build_list_includes_external_versions(self):
             reverse('builds_project_list', args=[self.pip.slug]),
         )
         self.assertEqual(response.status_code, 200)
-
+        
         self.assertIn(external_version_build, response.context['build_qs'])
+
+
+class HomepageViewTests(TestCase):
+
+    def test_projects_count(self):
+        url = reverse('homepage') 
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
```

Comment:
The tests can be improved.
You can make some dummy users and projects, then mark some users as `banned` and then check the value of `projects_count` to check if the projects with banned users are excluded.
We are using [Django Dynamic Fixture](https://django-dynamic-fixture.readthedocs.io/en/latest/) to make dummy data.

You can find many examples of that in our tests:
https://github.com/readthedocs/readthedocs.org/blob/356d811b740460afe1620435ece6c9993ba480a0/readthedocs/rtd_tests/tests/test_views.py#L203-L204

Reply:
you mean I have to replace queryset at L-306 to dummy data?

Reply:
Would be better to re-write the tests.

Reply:
@dojutsu-user ,Actually I'm not good in writting test cases, can you please help me with this?
as you said that i have to make dummy users and projects  using DDF ,
`self.user_profile = get(UserProfile, banned=False)
self.user = get(User, profile=user_profile)
self.project = get(Project, users=[user])`

Am i doing right ? :)

Reply:
You can write something like this:

```python3
class HomepageViewTests(TestCase):

    def setUp(self):
        self.homepage_url = reverse('homepage')

    def test_projects_count(self):
        user_1 = get(User)
        profile_1 = get(UserProfile, user=user_1, banned=True)

        user_2 = get(User)
        profile_2 = get(UserProfile, user=user_2, banned=True)

        user_3 = get(User)
        profile_3 = get(UserProfile, user=user_3, banned=False)

        project_a = get(Project, users=[user_1])
        project_b = get(Project, users=[user_2])
        project_c = get(Project, users=[user_3])

        spam_proj_count = Project.objects.filter(users__profile__banned=True).count()

        self.assertEqual(spam_proj_count, 2)
        self.assertEqual(User.objects.all().count(), 3)
        self.assertEqual(User.objects.filter(profile__banned=True).count(), 2)

        expected_count = Project.objects.all().count() - spam_proj_count
        response = self.client.get(self.homepage_url)

        self.assertEqual(response.status_code, 200)
        self.assertEqual(expected_count, response.context['projects_count'])
```

Reply:
Thank you @dojutsu-user